전남대 BE_김보민_2주차 과제 (2단계 ~ 3단계)

KBM05haruin30 opened the pull request:
### 1단계 리뷰 수정할 내용
- Controller에서 BindingReulst를 사용하지 않고 예외 처리하기
- 검증 메서드 추가하기

# 2단계
### 코드 작성하면서 어려웠던 점
인증을 위해 jwt 토큰을 사용해보려고 여러 자료를 찾아봤습니다. 하지만 대부분의 자료들은 SpringSecurity를 사용하고 있었고 저는 SpringSecurity를 모르고 있습니다. 그래서 SpringSecurity를 안쓰고 하는 방법을 찾는 것이 어려웠습니다. 그리고 지금 작성해 둔 코드가 제대로 작성된 것인지도 모르겠습니다...
제가 jwt를 사용해서 2단계 과제에 있는 secretKey를 사용했습니다. 이 secretKey를 JwtUtil.java에
`private static final String SECRET_KEY = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";`
이런 식으로 나타냈습니다. 다른 팀원에게 달린 리뷰를 보니 SECRET_KEY 같은 중요한 정보들은 노출이 되면 안된다고 달려있었습니다. 그래서 찾아보니 application-secret-key.properties에 추가해서 값을 불러오는 식으로 처리하는 것을 찾을 수 있었습니다. 한 번 따라해 보려고 application-secret-key.properties에 `secret_key=Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=` 라고 적었습니다. 그리고 application.properties에는 `spring.profiles.include=secret-key`를 적었습니다. 그 다음 secret_key를 사용하는 JwtUtil.java에 가서
`@Value("$secret_key}")
private String secretkey;`
이런 식으로 작성을 했습니다. 옳게 작성했다고 생각했는데 작동이 되지 않아 잘못된 점이 있는지 궁금합니다!

### 중점적으로 봐주셨으면 하는 것
코드가 잘 작성되었는지 궁금합니다.
이메일과 비밀 번호를 제대로 입력하면 DB에서 찾아 일치하면 토큰을 만들게 코드를 작성했습니다. 토큰을 만들었는데 이제 이 토큰을 어떻게 사용해야 하는지 모르겠습니다.
stpe2에 힌트로 Basic 인증이 있는데 제가 이해하는 것이 맞는지 모르겠습니다.
1. 사용자가 로그인을 하기 위해 email과 비밀 번호를 입력하면 로그인 페이지는 Basic base64로 인코딩하여 서버에 제공합니다.
2. 서버는 인코딩된 것을 제공 받고 그것을 디코딩하여 email과 비밀 번호를 얻어 DB에서 비교한 후 JWT 토큰을 생성합니다.
3. 이 생성된 토큰을 사용자에게 제공합니다.

이렇게 흘러가는 것이 맞나요?
맞다면 로그인 페이지에서 어떻게 입력한 email과 비밀 번호를 서버에 도착하기 전에 Basic base64로 인코딩하는지 궁금합니다.

응답 코드 관련해서도 궁금한 점이 있습니다. 아직 BindingResult를 사용해서 코드를 작성했습니다. 그래서 GlobalExceptionHandler에서는 이미 MethodArgumentNotValidException과 IllegalArgumentException을 사용 중입니다.MethodArgumentNotValidException을 @ResponseStatus를 사용해 BAD_REQUEST로 처리했습니다. 과제에서는 401 Unauthorized와 403 Forbidden을 반환하는 것으로 나와 있습니다. 
`@ResponseStatus(HttpStatus.Forbidden)
if (user == null || !user.password().equals(loginDTO.password())) {
            model.addAttribute("loginError", "잘못된 이메일 또는 비밀번호입니다.");
            return "login_user_form";
        }`
이런 식으로 사용이 불가능하다고 알고 있습니다. 이미 MethodArgumentNotValidException과 IllegalArgumentException을 사용 중인데 어떻게 해야 하는지 궁금합니다.

---
# 3단계
### 코드 작성하면서 어려웠던 점
클라이언트에게 토큰을 저장하고, 클라이언트가 토큰을 주기 위해서 '자바 스크립트'를 사용한다는 것을 알게 되었습니다. 자바 스크립트를 사용해본 적이 없어 찾아 보는데 시간이 많이 걸린 것 같습니다. 또한, 힌트로 주신 HandlerMethodArgumentResolver가 많이 어려웠습니다. 처음 보는 것이기도 하고 어떻게 사용해야 하는지 다른 사람들이 작성해둔 코드를 보거나 검색도 해보면서 했음에도 불구하고 어려웠습니다. 작성한 메소드를 사용하기 위해서 WebConfig를 만들어야 하는 것도 처음 알게 되었습니다. WebConfig와 HandlerMethodArgumentResolver에 관련된 자료가 있다면 공유 받고 싶습니다!
이번 단계도 처음 보는 것이 많다 보니 어렵다고 느껴졌습니다. 열심히 공부해서 쫓아가도록 하겠습니다.

### 중점적으로 봐주셨으면 하는 것
코드가 잘 작성되었는지 궁금합니다. 
어노테이션을 알맞게 만들었는지, LoginUserArgumentResolver을 알맞게 만들었는지 봐주셨으면 좋겠습니다. 다른 사람들의 리뷰를 보면 request와 response dto를 작성하라고 쓰여 있던데 왜 나누는지도 궁금합니다. 자바 스크립트를 사용한 적이 거의 없는데 잘 사용한 것인지, 개선할 점은 무엇인지 알고 싶습니다. 지금은 html 안에 <script>로 자바 스크립트를 사용 중입니다. html에서 자바 스크립트를 분리해서 사용하는 방법이 있을 것인데 어떤 방법이 더 나은지 궁금합니다. 자바 스크립트를 사용하지 않고 토큰과 Authorization의 유형을 전송하는 방법이 있는지 궁금합니다.

KBM05haruin30 committed:
docs : 2단계 - 회원 로그인 구현할 기능 목록 정리

KBM05haruin30 committed:
feat : jjwt 라이브러리 추가

KBM05haruin30 committed:
feat : User 객체 생성
필드는 id, name, email, password, role을 사용했습니다.

KBM05haruin30 committed:
feat : User DB 생성
User DB를 사용하기 위해 schema.sql과 data.sql에 코드를 추가했습니다. 역할을 3가지로 한 이유는 일반 사용자, 관리자, '카카오'키워드 사용 가능한 사용자 이렇게 생각했기 때문입니다.

KBM05haruin30 committed:
feat : UserDTO 생성
회원 가입을 할 때 User가 입력해야 하는 것은 이름, 이메일, 비밀 번호라고 생각했기 때문에 이렇게 작성했습니다.

KBM05haruin30 committed:
feat : LoginDTO 생성
로그인을 할 때는 이메일과 비밀 번호만 필요하다고 생각하여 이렇게 작성했습니다.

KBM05haruin30 committed:
feat : UserRepository 생성
회원 가입 시 User의 정보를 DB에 저장할 수 있도록 작성했습니다. 로그인 시 email을 기준으로 DB에서 정보를 찾을 수 있도록 작성했습니다.

KBM05haruin30 committed:
feat : JwtUtil 생성
JWT 토큰을 생성하고 검증하는 메서드를 작성했습니다. SECRET_KEY는 과제에 나와 있는 키를 그대로 사용했습니다.

KBM05haruin30 committed:
docs : 2단계 - 회원 로그인 구현할 기능 목록 수정

KBM05haruin30 committed:
feat : 회원 가입

KBM05haruin30 committed:
feat : 로그인 및 로그인 성공 시 토큰 생성

KBM05haruin30 committed:
feat :회원 가입 화면 만들기

KBM05haruin30 committed:
feat : 로그인 화면 만들기

KBM05haruin30 committed:
feat : 회원 가입 성공 화면 만들기

KBM05haruin30 committed:
feat : 로그인 성공 화면 만들기
토큰이 제대로 나오는지 궁금하여 로그인 성공 화면에 토큰을 출력하게 했습니다. 아직 로그인이 완료된 후 어디로 이동할지는 정하지 못했습니다.

KBM05haruin30 committed:
refactor : 코딩 컨벤션 준수를 위한 코드 정리

KBM05haruin30 committed:
feat : User 객체에서 검증

KBM05haruin30 committed:
feat : UserService 생성
UserService 생성합니다. 그거에 맞게 UserController를 수정해줍니다. UserDTO를 User로 변환 하는 것이므로 Service를 만들 후 여기서 변환 시켜 줍니다.

KBM05haruin30 committed:
feat : 검증 메서드 추가

KBM05haruin30 committed:
feat : 검증 메서드 제거
검증 메서드가 잘 작동하는지 코드를 실행시켰는데 입력 받는 순간 멈춰버립니다. 잘못 작성한 메서드라고 생각해서 제거했습니다.

KBM05haruin30 committed:
feat : 로그인 성공 시 상품 목록으로 이동

KBM05haruin30 committed:
docs : 3단계 - 위시 리스트 구현할 기능 목록 정리

KBM05haruin30 committed:
feat : Wishlist 객체 생성

KBM05haruin30 committed:
feat : Wishlist DB 생성

KBM05haruin30 committed:
feat : WishlistRequest 생성

KBM05haruin30 committed:
feat : WishlistRepository 생성
위시 리스트 물품 조회, 삭제, 추가

KBM05haruin30 committed:
feat : WishlistService 생성
위시 리스트 물품 조회, 삭제, 추가

KBM05haruin30 committed:
feat : LoginUser 어노테이션 생성

KBM05haruin30 committed:
feat : WishlistController 생성
위시 리스트 물품 조회, 삭제, 추가

KBM05haruin30 committed:
feat : wishlist html 생성

KBM05haruin30 committed:
feat : LoginUserArgumentResolver 생성

KBM05haruin30 committed:
feat : WebConfig 생성

KBM05haruin30 committed:
feat : jwt 사용을 위한 html 수정
localStorage로 클라이언트가 토큰을 저장하게 합니다. 추가로 로그인 시 토큰을 보여주지 않게 수정했습니다.

KBM05haruin30 committed:
feat : 위시 리스트로 이동 및 추가 기능

KBM05haruin30 committed:
feat : 위시 리스트에서 삭제 기능

KBM05haruin30 committed:
refactor : 코드 정리

KBM05haruin30 committed:
refactor : dto 패키지 생성
dto 패키지 생성으로 인한 코드 수정

KBM05haruin30 committed:
refactor : 코딩 컨벤션 준수

KBM05haruin30 committed:
refactor : dto 패키지 위치 수정

Hongdonggeon commented on src/main/java/gift/controller/GlobalExceptionHandler.java:
```diff
@@ -25,4 +25,12 @@ public Map<String, String> handleMethodArgumentNotValidException(
+        error.put("error", ex.getMessage());
+        return error;
+    }
+
```
`@Valid` 에러인경우 -> MethodArgumentNotValidException.class 인 경우에도 추가해주면 좋을 것 같아요

KBM05haruin30 replied:
위에
`@ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public Map<String, String> handleMethodArgumentNotValidException(
        MethodArgumentNotValidException ex) {
        Map<String, String> errors = new HashMap<>();
        ex.getBindingResult().getAllErrors().forEach((error) -> {
            String fieldName = ((FieldError) error).getField();
            String errorMessage = error.getDefaultMessage();
            errors.put(fieldName, errorMessage);
        });`
        를 작성했었습니다. 멘토님 말씀은 밑의 예외의
        `@ExceptionHandler(IllegalArgumentException.class)` 부분을
        `@ExceptionHandler(IllegalArgumentException.class, MethodArgumentNotValidException.class)`로 수정하는 것이 좋을 것 같다는 것인가요??

Hongdonggeon replied:
아니요 MethodArgumentNotValidException을 추가해주면 좋을 것 같다는 말이었어요!

KBM05haruin30 replied:
넵 알겠습니다!!

Hongdonggeon commented on src/main/java/gift/resolver/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,55 @@
+    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
+        NativeWebRequest webRequest,
+        org.springframework.web.bind.support.WebDataBinderFactory binderFactory) throws Exception {
+        String authorizationHeader = webRequest.getHeader("Authorization");
```
헤더명은 상수로 관리해도 좋을 것 같아요

KBM05haruin30 replied:
넵! 헤더 명을 상수로 변경했습니다. 추가로 "Bearer "도 상수로 관리하면 좋을 것 같아 AUTHORIZATION_NAME으로 만들었습니다.

Hongdonggeon commented on src/main/java/gift/util/JwtUtil.java:
```diff
@@ -0,0 +1,36 @@
+
+public class JwtUtil {
+
+    private static final String SECRET_KEY = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
시크릿키를 이렇게 공개적으로 노출하면 보안적으로 취약할것같아요! 
application.yml 이외에 다른 yml파일을 만들어서 노출하지 않도록 관리하면 좋을 것 같아요

KBM05haruin30 replied:
넵! 알겠습니다!

KBM05haruin30 replied:
application-secret-key.properties를 만들고 여기에 시크릿 키를 넣었습니다. Application에는 @PropertySource를 사용하여 application.properties 파일 외에 application-secret-key.properties를 로드할 수 있었습니다.
@Valid를 사용하여 시크릿 키를 불러왔습니다. @Valid는 static에 사용할 수 없다고 하여 static을 지운 후 JwtUtil을 사용하고 있는 파일에 생성자에 추가하여 사용할 수 있게 했습니다. 또한 @Component를 사용하지 않았을 때 bean 관련 에러가 나온 것을 확인해서 JwtUtil에 @Component를 사용했습니다.

Hongdonggeon replied:
좋습니다!

KBM05haruin30 committed:
feat : saveToken 실행
페이지가 로드되면 saveToken을 실행하도록 합니다.

Hongdonggeon commented on src/main/resources/templates/login_user_form.html:
```diff
@@ -0,0 +1,26 @@
+<!DOCTYPE html>
```
로그인 화면에 회원가입할 수 있는 버튼을 만들면 좋을 것 같아요

KBM05haruin30 replied:
넵! 로그인 화면에서 회원 가입 화면으로 이동할 수 있는 버튼을 만들었습니다. 추가로 회원 가입 화면에서 로그인 화면으로 이동하는 버튼, 상품 목록 화면에서 로그인 화면으로 이동할 수 있는 버튼도 만들었습니다.

Hongdonggeon commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -23,7 +24,7 @@ public Product findProductsById(Long id) {
 
     public void saveProduct(ProductDTO productDTO) {
-        productRepository.saveProduct(toEntity(productDTO,null));
+        productRepository.saveProduct(toEntity(productDTO, null));
```
toEntity 메서드를 private으로 선언해도 될 것 같아요

KBM05haruin30 replied:
넵! 같은 클래스 안에서만 작동하므로 private로 하는 것이 더 맞는 것 같습니다! 바로 수정하겠습니다!

Hongdonggeon commented on src/main/java/gift/controller/AdminController.java:
```diff
@@ -1,7 +1,8 @@
+import gift.service.ProductService;
 import jakarta.validation.Valid;
 import java.util.List;
 import org.springframework.stereotype.Controller;
```
로그인을 이용하지 않고 상품 추가, 삭제를 할 수 있네요

KBM05haruin30 replied:
넵! 다음과 같은 순서로 작동하도록 수정했습니다.
1. 로그인을 성공하면 토큰을 발급 받습니다.
2. 토큰이 있는 경우에만 상품 목록에서 상품 추가, 상품 수정, 상품 삭제 버튼이 보이도록 수정했습니다.

Hongdonggeon commented on src/main/resources/templates/wishlist.html:
```diff
@@ -0,0 +1,61 @@
+  </script>
+</head>
+<body>
+<h1>위시리스트</h1>
```
위시리스트 화면으로 로그인 성공해도 진입할 수가 없네요

KBM05haruin30 replied:
지금은 로그인에 성공한 후 위시리스트에 진입할 수 있도록 수정 완료했습니다!
'feat : html의 id 사용' 에 관하여 궁금한 점이 있습니다.
현재 저는 다음과 같이 코드를 작성했습니다.
const token = document.getElementById("token").innerText;
이렇게 작성하다 보니 문제점이 생겼습니다. 바로 이 html에서 밑의 코드처럼 작성을 해주어야 한다는 것입니다.
<p id="token" th:text="${token}" style="display: none"></p>
display:none이라는 속성을 통해 html에서는 보이지 않지만 브라우저의 개발자 도구를 통해 토큰이 무엇인지 볼 수 있었습니다. 이렇게 저장하는 것이 아닌 서버에서 보낸 토큰을 바로 저장할 수 있는 방법이 궁금합니다.

Hongdonggeon replied:
저도 프론트는 지식이 얕아서.. 한번 찾아보고 답변드리겠습니다!  죄송해요 ㅠ

KBM05haruin30 replied:
넵 감사합니다!

Hongdonggeon reviewed (changes requested):
안녕하세요 보민님 
미션 구현하시느라 수고하셨습니다!! 
보민님 어플리케이션 실행해봤을 때 요구사항에 조금 부족한 점이 2개 있는 것 같아서요! 
- 현재 로그인에 성공하더라도 위시리스트에 진입할 수 가 없어요
- 상품 추가 및 삭제는 로그인 여부에 상관없이 진행할 수 있는 것 같아요
그리고 화면에서 상품 목록에서 로그인할 수 있는 버튼이 있으면 좋을 것 같고, 로그인 화면에서 회원가입으로 갈 수 있는 버튼이 있으면 좋을 것 같아요

질문 주신 것들 정리해서 코멘트에 올려놓을게요! 수고하셨어요!

KBM05haruin30 committed:
feat : html의 id 사용
id를 사용하지 않으면 token이 저장되지 않는 것을 확인했습니다.

KBM05haruin30 committed:
feat : toEntity 메서드 private로 변경

KBM05haruin30 commented:
지금은 로그인에 성공한 후 위시리스트에 진입할 수 있도록 수정 완료했습니다!
'feat : html의 id 사용' 에 관하여 궁금한 점이 있습니다. 
현재 저는 다음과 같이 코드를 작성했습니다.
`const token = document.getElementById("token").innerText;`
이렇게 작성하다 보니 문제점이 생겼습니다. 바로 이 html에서 밑의 코드처럼 작성을 해주어야 한다는 것입니다.
`<p id="token" th:text="${token}" style="display: none"></p>`
display:none이라는 속성을 통해 html에서는 보이지 않지만 브라우저의 개발자 도구를 통해 토큰이 무엇인지 볼 수 있었습니다. 이렇게 저장하는 것이 아닌 서버에서 보낸 토큰을 바로 저장할 수 있는 방법이 궁금합니다.

KBM05haruin30 committed:
feat : Header_name, Authorization_name 상수로 변환

KBM05haruin30 committed:
feat : 회원 가입 버튼
로그인 화면에서 회원 가입 화면으로 이동하는 버튼을 만들었습니다.

KBM05haruin30 committed:
feat : 로그인 버튼
회원 가입 화면에서 로그인 화면으로 이동하는 버튼을 만들었습니다.

KBM05haruin30 committed:
feat : 로그인 버튼
상품 목록 화면에서 로그인 화면으로 이동하는 버튼을 만들었습니다.

KBM05haruin30 committed:
feat : 로그인일 때 상품 추가, 삭제, 수정 가능
로그인을 하면 토큰이 생성됩니다. 토큰이 존재한다면 상품 추가, 삭제, 수정이 가능하게 했습니다. 토큰이 존재하지 않으면 버튼이 보이지 않게 수정했습니다.

KBM05haruin30 committed:
feat : 시크릿 키 숨기기

Hongdonggeon reviewed (approved):
안녕하세요 보민님😃
리뷰 반영해주신 것 잘 확인했습니다! 
학습하면서 성장하시는 것도 보기 좋습니다 :)
한가지만 더 코멘트 드릴게요! 
현재 어플리케이션에서 로그인을 해도 로그인 버튼이 계속 노출되는데 이 부분 개선하면 좋을 것 같아요
3주차 과제 진행하면서 수정해주시면 좋을 것 같아요 
수고하셨습니다! 머지할게요~
